### PR TITLE
Feature/interpolator

### PR DESCRIPTION
**Pull request by vhuber**
linked to #733

### CODE REVIEW

**Comment by prudhomm** [comment]
is it done @vhuber ?

**Comment by prudhomm** [comment]
@vhuber 
 - [ ] has the testsuite been re-run ?
 - [ ] is there a test for this feature ?

if not could you post here the result of the testsuite including the test for this feature ? 
thank you in advance

**Comment by vhuber** [comment]
@prudhomm 
```sh
95% tests passed, 16 tests failed out of 303

Label Time Summary:
testalg              =  75.27 sec
testcore             =  57.20 sec
testcrb              = 633.60 sec
testdiscr            = 2143.06 sec
testfilters          = 218.32 sec
testfit              =   9.94 sec
testintegration      = 151.84 sec
testinterpolation    = 730.18 sec
testleaks            =  93.22 sec
testls               =  14.03 sec
testmaterial         =   4.83 sec
testmath             =   3.70 sec
testmesh             =  79.75 sec
testmodels           =   5.87 sec
testopt              =   4.34 sec
testpoly             = 272.28 sec
testts               =  20.01 sec
testvf               = 1088.12 sec

Total Test time (real) = 571.33 sec

The following tests FAILED:
	 74 - feelpp_test_concatenate-np-6 (Failed)
	 75 - feelpp_test_concatenate-np-1 (Failed)
	144 - feelpp_test_integration_relatedmesh-np-1 (Failed)
	205 - feelpp_test_submesh-np-1 (Timeout)
	214 - feelpp_test_matching-np-1 (Failed)
	216 - feelpp_test_mortar-np-1 (Failed)
	217 - feelpp_test_mortar2-np-1 (Failed)
	237 - feelpp_test_eim-np-1 (Failed)
	238 - feelpp_test_eim_grepl-np-1 (Failed)
	239 - feelpp_test_eim_composite-np-1 (Failed)
	240 - feelpp_test_rbspace-np-6 (Timeout)
	242 - feelpp_test_rbspacevector-np-6 (Timeout)
	248 - feelpp_test_bdf3-np-1 (Failed)
	269 - feelpp_test_convolve-np-6 (Timeout)
	300 - feelpp_test_nofeel_interpolator-np-6 (Failed)
	301 - feelpp_test_nofeel_interpolator-np-1 (Failed)
Errors while running CTest
```

**Comment by vhuber** [comment]
@prudhomm , I guess it can be merged.

File: feel/feelfit/fit.hpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include "interpolator.hpp"
+
+#ifndef __FEELPP_FIT_H
```
Comment (prudhomm) [inline feel/feelfit/fit.hpp]: use FEELPP_FIT_HPP

File: feel/feelfit/fit.hpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    
+    typedef typename expression_type::value_type value_type;
+    // scalar, vectorial (to be checked)
+    typedef typename expression_type::evaluate_type evaluate_type;
+
+    typedef Fit<ExprT> this_type;
+
+    explicit Fit( expression_type const & __expr, std::string dataFile = soption("fit.datafile"), int T = 3 ) : M_expr(__expr)
+    {
+        M_interpolator = Interpolator::New(static_cast<interpol_type>(T), dataFile);
```
Comment (prudhomm) [inline feel/feelfit/fit.hpp]: how about deleting the pointer ? there is a memory leak here.

File: feel/feelfit/fit.hpp [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    }
+    Fit(Fit const & te) : M_expr(te.M_expr),M_interpolator(te.M_interpolator) {}
+    //~Fit(){}
+    expression_type const& expression() const
+    {
+          return M_expr;
+    }
+    Interpolator* const& interpolator() const
```
Comment (prudhomm) [inline feel/feelfit/fit.hpp]: why return a pointer here and not the object itself ?

File: feel/feelfit/fit.hpp (hunk 1 above)
Comment (vhuber) [inline feel/feelfit/fit.hpp]: Done.

File: feel/feelfit/fit.hpp [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+    private:
+        tensor_expr_type M_tensor_expr;
+        Interpolator* M_interpolator;
+    };
+    /// end of tensor
+    private:
+        mutable expression_type  M_expr;
+        Interpolator*  M_interpolator;
```
Comment (prudhomm) [inline feel/feelfit/fit.hpp]: could/should be a `std::unique_ptr`

File: feel/feelfit/interpolator.hpp [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+        { 
+            double y1 = 0.;
+            double y2 = 0.;
+            double x1 = 0.;
+            double x2 = 0.;
+            getCoef(_x, y1, y2, x1, x2);
+            return ((y1-y2))/(x1-x2);
+        }
+    private:
+        void getCoef(value_type _x,
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: function too big, should outside for reability

File: feel/feelfit/interpolator.hpp [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+        /*
+         * natural : f''(x0) = f''(xn) = 0
+         * clamped : f'(x0) = 0; f'(xn) = 0
+         */
+        enum extrapol_type_spline
+        {
+            natural=0,
+            clamped
+        };
+        InterpolatorSpline( std::vector<pair_type> data, extrapol_type_spline r = natural, extrapol_type_spline l = natural ) : super(data), M_r_type(r), M_l_type(l)
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: function too big, should go outside

File: feel/feelfit/interpolator.hpp [hunk 7]
```diff
@@ -0,0 +1,625 @@
...
+            return 3.*a*_x*_x+2.*b*_x+c;
+        }
+
+    private:
+        extrapol_type_spline M_r_type;
+        extrapol_type_spline M_l_type;
+
+        Eigen::VectorXd sol;
+
+        void getCoef(value_type _x,
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: please use full name and no `get` (see feel++ coding style)

File: feel/feelfit/interpolator.hpp [hunk 8]
```diff
@@ -0,0 +1,625 @@
...
+ */
+class InterpolatorAkima : public Interpolator
+{
+    public:
+        typedef Interpolator super;
+        using value_type = double;
+        using pair_type = std::pair<value_type, value_type>;
+        using SpMat = Eigen::SparseMatrix<value_type>; 
+        using T = Eigen::Triplet<value_type>;
+        InterpolatorAkima( std::vector<pair_type> data) : super(data)
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: too big, outside

File: feel/feelfit/interpolator.hpp [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+        {
+            double a = 0.;
+            double b = 0.;
+            double c = 0.;
+            double d = 0.;
+            getCoef(_x, a, b, c, d);
+            return 3.*a*_x*_x+2.*b*_x+c;
+        }
+    private:
+        value_type sp(int i)
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: what does it do ? not explict name

Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: and too  big

File: feel/feelfit/interpolator.hpp [hunk 10]
```diff
@@ -0,0 +1,625 @@
...
+            else if(j > M_data.size()-2 )
+                return 2*d(j-1)-d(j-2);
+            else
+            {
+                return (M_data.at(j+1).second - M_data.at(j).second) /  (M_data.at(j+1).first - M_data.at(j).first);
+            }
+        }
+        Eigen::VectorXd sol;
+
+        void getCoef(value_type _x,
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: see above

File: feel/feelfit/interpolator.hpp [hunk 11]
```diff
@@ -0,0 +1,625 @@
...
+                a = sol(index*4+0);
+                b = sol(index*4+1);
+                c = sol(index*4+2);
+                d = sol(index*4+3);
+            }
+        }
+
+}; // InterpolatorSpline
+
+Interpolator* Interpolator::New( interpol_type type, std::vector<pair_type> data)
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: return std::unique_ptr

File: feel/feelfit/fitdiff.hpp [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+    {
+        M_interpolator = Interpolator::New(static_cast<interpol_type>(T), dataFile);
+    }
+    FitDiff(FitDiff const & te) : M_expr(te.M_expr),M_interpolator(te.M_interpolator) {}
+    //~FitDiff(){}
+    expression_type const& expression() const
+    {
+          return M_expr;
+    }
+    Interpolator* const& interpolator() const
```
Comment (prudhomm) [inline feel/feelfit/fitdiff.hpp]: see `fit`

File: feel/feelfit/fitdiff.hpp [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+        Interpolator* M_interpolator;
+    };
+    /// end of tensor
+        private:
+            mutable expression_type  M_expr;
+            Interpolator*  M_interpolator;
+    };
+
+/**
+ * \brief FitDiff
```
Comment (prudhomm) [inline feel/feelfit/fitdiff.hpp]: what does it do ? 
please, add some doc.
what do you differentiate against ? what are the variables ?

Comment (prudhomm) [inline feel/feelfit/fitdiff.hpp]: actually \brief is not really necessary if you have indeed a first line with a brief statement

File: feel/feelfit/interpolator.hpp [hunk 14]
```diff
@@ -0,0 +1,625 @@
...
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#include <feel/feel.hpp>
+#include <feel/feelalg/backendeigen.hpp>
+
+#ifndef __FEELPP_INTERPOLATOR_H
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: FEELPP_INTERPOLATOR_HPP or FEELPP_FIT_INTERPOLATOR_HPP

File: feel/feelfit/interpolator.hpp [hunk 15]
```diff
@@ -0,0 +1,625 @@
...
+#define __FEELPP_INTERPOLATOR_H 1
+
+enum interpol_type
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: should go in feelfit/enums.hpp and have a name `InterpolationType`

File: feel/feelfit/interpolator.hpp [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+{
+    P0=0,
+    P1,
+    Spline,
+    Akima
+};
+
+/*
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: doxygen doc

File: feel/feelfit/interpolator.hpp [hunk 17]
```diff
@@ -0,0 +1,625 @@
...
+        virtual value_type diff(double _x)
+        {
+            LOG(INFO) << "[interpolator] Base class diff()\n";
+            return 0.;
+        }
+    protected:
+        std::vector<std::pair<value_type, value_type>> M_data;
+};// class interpolBase
+
+class InterpolatorP0 : public Interpolator
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: Doxygen doc

File: feel/feelfit/interpolator.hpp [hunk 18]
```diff
@@ -0,0 +1,625 @@
...
+        enum interpol_type_P0
+        {
+            left=0,
+            right,
+            center
+        };
+        InterpolatorP0( std::vector<pair_type> data, interpol_type_P0 iType = left ) : super(data), M_iType(iType) {}
+        InterpolatorP0(const InterpolatorP0 &) = default;
+        value_type diff(double _x) { return 0.; }
+        value_type operator()(double _x)
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: too big

File: feel/feelfit/interpolator.hpp [hunk 19]
```diff
@@ -0,0 +1,625 @@
...
+    public:
+        typedef Interpolator super;
+        using value_type = double;
+        using pair_type = std::pair<value_type, value_type>;
+        /*
+         * zero : P1(x ! \in{x_0, x_n} ) = 0
+         * constant : P1(x = x0) = x0 || P1(x = xn) = xn
+         * extrapol : linear extrapolation
+         */
+        enum extrapol_type_P1
```
Comment (prudhomm) [inline feel/feelfit/interpolator.hpp]: should go in feelfit/enums.hpp

**Comment by prudhomm** [comment]
it seems that there are also some indentation issues with respect to feel++ coding standard

File: feel/feelfit/fit.hpp (hunk 4 above)
Comment (vhuber) [inline feel/feelfit/fit.hpp]: @prudhomm Is there a reason to have the Interpolator in the class and in the class.struct ?

**Comment by vhuber** [comment]
@prudhomm Merge ?

**Comment by vhuber** [comment]
@prudhomm Up

**Comment by vhuber** [comment]
@prudhomm Up.

**Comment by prudhomm** [comment]
@vhuber it seems the branch is conflicting with develop. could you have a look before the merge please ? thanks in advance

**Comment by vhuber** [comment]
@prudhomm I guess you can merge now.

**Comment by prudhomm** [comment]
thanks @vhuber  👍

**Pull request merged**